File: leetcode/algorithms/530_minimum_absolute_difference_in_bst/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var getMinimumDifference = function (root) {
  const array = [];
  let result = 10 ** 5 + 1;

  /**
   * @param {TreeNode} node
   */
  const bstToArray = (node) => {
    if (node) {
      bstToArray(node.left);
      array.push(node.val);
      bstToArray(node.right);
    }
  };

  bstToArray(root);

  for (let i = 1; i < array.length; i++) {
    const diff = Math.abs(array[i - 1] - array[i]);
    if (diff < result) {
      result = diff;
    }
  }

  return result;
};


// Best Solution
var bestSolution = function (root) {
  // Converting tree to an sorted array
  const arr = toArray(root);
  let out = Infinity;
  let i = 1;
  // Now, simply finding the difference
  while (i < arr.length) {
    let diff = Math.abs(arr[i - 1] - arr[i]);
    if (diff < out) {
      out = diff;
    }
    i++;
  }
  return out;
};

function toArray(root, out = []) {
  if (root) {
    toArray(root.left, out);
    out.push(root.val);
    toArray(root.right, out);
    return out;
  }
}

File: leetcode/algorithms/530_minimum_absolute_difference_in_bst/main.dart
class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class MinimumAbsoluteDifferenceInBST {
  int min = 100000;
  int val = 100000;

  int getMinimumDifference(TreeNode? root) {
    if (root != null) {
      getMinimumDifference(root.left);

      int diff = (root.val - val).abs();
      if (diff < min) {
        min = diff;
      }
      val = root.val;

      getMinimumDifference(root.right);
    }
    return min;
  }

  // Solution
  double prev = double.negativeInfinity;
  double ans = double.infinity;

  int solution(TreeNode? root) {
    inOrder(root);
    return ans.toInt();
  }

  void inOrder(TreeNode? root) {
    if (root == null) return;

    inOrder(root.left);

    if (prev != double.negativeInfinity) {
      ans = ans < (root.val - prev).abs() ? ans : (root.val - prev).abs();
    }
    prev = root.val.toDouble();

    inOrder(root.right);
  }
}
